### PR TITLE
[AUTH-1236] Prevent conditional requests from being made during tape recording, so that conditional responses are not returned during replay

### PR DESCRIPTION
**Pull request by timdawborn**
Prevent conditional requests from being made during tape recording, so that conditional responses are not returned during replay.

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -57,6 +60,15 @@ export class RecordReplayServer {
         return;
       }
 
+      if (
+        this.preventConditionalRequests &&
+        (req.method === "GET" || req.method === "HEAD")
+      ) {
+        // Headers are always coming in as lowercase.
+        delete req.headers["if-modified-since"];
+        delete req.headers["if-none-match"];
```
Comment (rickyrattlesnake) [inline src/server.ts:69]: Question: do we need to remove the headers from https://nodejs.org/api/http.html#messagerawheaders as well? I assume not if we're only forwarding `req.headers` elsewhere.

**Review by rickyrattlesnake — APPROVED** [review]

File: src/server.ts (hunk 1 above)
Comment (timdawborn) [reply src/server.ts:69]: Good question. I don't believe so as we're copying the headers out of `req` to pass upstream later on.

File: src/cli.ts [hunk 2]
```diff
@@ -24,6 +24,11 @@ async function main(argv: string[]) {
       "Request headers to redact",
       commaSeparatedList
     )
+    .option(
+      "--prevent-conditional-requests <flag>",
+      "When running in record mode, remove `If-*` headers from outgoing requests in an attempt to prevent the suite of conditional responses being returned (304).",
+      "true"
+    )
```
Comment (rickyrattlesnake) [inline src/cli.ts:31]: nit: perhaps a negated boolean flag is a nicer CLI interface https://github.com/tj/commander.js/#other-option-types-negatable-boolean-and-booleanvalue? 

`--no-drop-conditional-request-headers` ?

**Review by rickyrattlesnake — APPROVED** [review]

File: src/cli.ts (hunk 2 above)
Comment (timdawborn) [reply src/cli.ts:31]: Ah sure, nice idea. I didn't bother looking up the docs for this command line arg parsing library to see if there was an option for that.

File: src/cli.ts [hunk 3]
```diff
@@ -55,6 +62,12 @@ async function main(argv: string[]) {
     );
   }
 
+  if (initialMode === "record" && preventConditionalRequests) {
+    console.info(
+      "The prevent conditional requests flag is enabled in record mode. All received `If-*` headers will not be forwarded on upstream and will not be recorded in tapes."
```
Comment (mcinbell) [inline src/cli.ts:67]: 👍🏼

**Review by mcinbell — APPROVED** [review]
I like @rickyrattlesnake's suggestion of boolean flag, but other than that it looks good to me.

**Pull request merged**